banners: Redesign navbar banners.

sayamsamal opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #31280

Related CZO discussions: [[1]](https://chat.zulip.org/#narrow/channel/431-redesign-project/topic/top.20banners/near/1925940), [[2]](https://chat.zulip.org/#narrow/channel/101-design/topic/banner.20redesign/near/1987506)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

## Screenshots and screen captures
<details>
<summary>DESKTOP NOTIFICATIONS BANNER</summary>

![notifications_banner_sm](https://github.com/user-attachments/assets/b4e35676-7f72-4a5c-8daa-76a899172b05)
![notifications_banner_md](https://github.com/user-attachments/assets/fa4fdcb3-d292-47be-9fdf-dcc3207a6f47)
![notifications_banner_lg](https://github.com/user-attachments/assets/b10e0440-afd4-40b2-8897-b8b5d2e2e1db)

</details>

<details>
<summary>CONFIGURE OUTGOING MAIL BANNER</summary>

![configure_email_banner_sm](https://github.com/user-attachments/assets/071d9bb9-66e5-44b3-9f2a-8eea4dfc6402)
![configure_email_banner_md](https://github.com/user-attachments/assets/e9bf7391-a7dc-4d61-a2f4-283e9f910124)
![configure_email_banner_lg](https://github.com/user-attachments/assets/e59a98cb-710a-4c05-a4f7-7dc87f444820)

</details>

<details>
<summary>PROFILE MISSING REQUIRED FIELDS BANNER</summary>

![prof_missing_banner_sm](https://github.com/user-attachments/assets/38aba6bb-a850-45e5-b876-fea473c05e19)
![prof_missing_banner_md](https://github.com/user-attachments/assets/fc95374c-52f2-4573-8f2d-34538023abb5)
![prof_missing_banner_lg](https://github.com/user-attachments/assets/aba0dc0b-6ba4-4e97-8de4-82a38997c7de)

</details>

<details>
<summary>ORGANIZATION PROFILE INCOMPLETE BANNER</summary>

![org_prof_banner_sm](https://github.com/user-attachments/assets/3ffc749e-e800-41f5-b301-7cdb7e6cc546)
![org_prof_banner_md](https://github.com/user-attachments/assets/4cff5e5f-43b2-4d2c-984c-626629e2c8ff)
![org_prof_banner_lg](https://github.com/user-attachments/assets/d0a1c704-29bd-4d36-9c13-8f5b85574f25)


</details>

<details>
<summary>SERVER NEEDS UPGRADE BANNER</summary>

![server_update_banner_sm](https://github.com/user-attachments/assets/de8bc5e4-01f5-4ca8-8870-29e062cb22f0)
![server_update_banner_md](https://github.com/user-attachments/assets/def2845d-bb0c-4afc-bed9-f713c95a53ce)
![server_update_banner_lg](https://github.com/user-attachments/assets/d33a0b38-3c64-4b86-b3bb-b95a3896b15f)


</details>

<details>
<summary>BANKRUPTCY BANNER</summary>

![bankruptcy_banner_sm](https://github.com/user-attachments/assets/8569d22c-69c8-456d-8b30-a4fb544161d3)
![bankruptcy_banner_md](https://github.com/user-attachments/assets/83b42ac3-c62f-45a5-9743-bd3e786dd6dd)
![bankruptcy_banner_lg](https://github.com/user-attachments/assets/6822234d-0a79-4b39-8739-5c03227fb443)

</details>

<details>
<summary>DEMO ORGANIZATION DEADLINE BANNER</summary>

![demo_org_banner_sm](https://github.com/user-attachments/assets/db02a5a3-02cb-4e5b-92b9-fa4fd063ff2a)
![demo_org_banner_md](https://github.com/user-attachments/assets/0b7761d8-4ae2-4d7f-b5dd-fcd48a631119)
![demo_org_banner_lg](https://github.com/user-attachments/assets/dae4da19-89e4-4a3a-a2e9-1cfbc05e19f3)


</details>

<details>
<summary>TIME ZONE UPDATE OFFER BANNER</summary>

![time_zone_banner_sm](https://github.com/user-attachments/assets/d209bebd-6792-4b54-b4c2-6bce1ade607e)
![time_zone_banner_md](https://github.com/user-attachments/assets/ef405be1-65bb-4a28-adc8-4178872a8d83)
![time_zone_banner_lg](https://github.com/user-attachments/assets/8ce2821b-db7c-41cf-accf-f2cfaec85651)


</details>

<details>
<summary>BANNER PROGRESS BAR</summary>

![Kapture 2025-01-27 at 03 33 41](https://github.com/user-attachments/assets/766fa3e3-dcc1-45ba-8946-7210a0241509)

</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Cool! Looks like the banner text isn't scaling properly with font size. At 20px:

![Screenshot 2025-01-22 at 08 37 29@2x](https://github.com/user-attachments/assets/04f931db-6028-444c-90ba-8c2c45f3410b)

alya commented:
Seeing this in the app, I still have the same feedback as in https://github.com/zulip/zulip/pull/32852#issuecomment-2572465617 about the buttons feeling too close to the text.

sayamsamal commented:
## Updates
- Increased the gap between the banner label and the actions buttons.
![Screenshot 2025-01-26 at 9 33 32 PM](https://github.com/user-attachments/assets/520537ba-b963-4621-a4ec-0d08def9f3d1)
- Fixed the banner label not scaling with font size.
![Screenshot 2025-01-27 at 3 29 59 AM](https://github.com/user-attachments/assets/c435bac7-4afd-4939-952b-53af3f2d537e)
- Added a temporary commit, to add an option in left sidebar to test all the navbar banners. We can remove this commit before merging the PR.
![Screenshot 2025-01-27 at 3 30 37 AM](https://github.com/user-attachments/assets/d66a7e46-e1c9-4b09-859f-a322f2f9dee0)
- Fixed the failing test cases.

sayamsamal commented:
@alya could you give this PR another look?

alya commented:
I love the left sidebar button for testing!

alya commented:
I see the button isn't vertically centered. Should it be?

![Screenshot 2025-01-27 at 08 22 14](https://github.com/user-attachments/assets/90c56c81-0def-4d7e-9dfc-a40429fd5016)

alya commented:
I pushed a commit to make the desktop notifications banner easier to understand -- please review and squash. I'll note it in #design as well in case folks have feedback.

alya commented:
[CZO thread](https://chat.zulip.org/#narrow/channel/101-design/topic/desktop.20notifications.20banner/near/2063442)

sayamsamal commented:
> I see the button isn't vertically centered. Should it be?
> 
> ![Screenshot 2025-01-27 at 08 23 55](https://private-user-images.githubusercontent.com/2090066/407000708-727dd036-2a1d-4acd-b5b6-9c3128897d20.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwNDkzMTcsIm5iZiI6MTczODA0OTAxNywicGF0aCI6Ii8yMDkwMDY2LzQwNzAwMDcwOC03MjdkZDAzNi0yYTFkLTRhY2QtYjViNi05YzMxMjg4OTdkMjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjhUMDcyMzM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGNlMWY0NjZlOWIxMTA0ZGMyNjZhMzdiOTFkMDU3OGU4ZDUwNmY2ZmJhZWYzNWEyMTg2Nzk3MTIzMDgwMjE3NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.iRbTs_yl4p-7cmrlt06wjpN0F-t0Sa-zy8Ym8qvGIyc)
> 
> Maybe the text is a bit higher than I'd expect too?

I checked. The button is actually vertically centered, but maybe the addition of the bottom border skews the perspective a bit? The text is also centrally placed in button, does the screenshot below with a descender in it feel better aligned than the typography with just the ascenders?
![Screenshot 2025-01-28 at 1 19 24 PM](https://github.com/user-attachments/assets/dd2785f1-d197-4b85-91c8-0b5e523fcd89)

sayamsamal commented:
Squashed the commit by @alya after reviewing it.

timabbott commented on web/src/banners.ts:
```diff
@@ -0,0 +1,50 @@
+    $("body").on("click", ".banner .banner_close_action", function (this: HTMLElement) {
+        const $banner = $(this).closest(".banner");
+        close($banner);
+    });
```
Should this be using `preventDefault` or `stopPropagation`? I'm not sure it matters; the main risk I see is the default behavior of moving focus to the X button.

sayamsamal replied:
Added. Made sense to add `e.preventDefault()` and `e.stopPropagation()` here since the event handler is registered on the body.

timabbott commented on web/src/navbar_alerts.ts:
```diff
@@ -162,6 +160,36 @@ export function should_offer_to_update_timezone(): boolean {
+    buttons: [
+        {
+            type: "primary",
+            label: "Enable notifications",
```
Labels not tagged for translation don't work.

sayamsamal replied:
Fixed. My bad, it was really obvious but I somehow overlooked it.

timabbott commented on web/src/banners.ts:
```diff
@@ -42,6 +42,32 @@ export function close($banner: JQuery): void {
+    let $progress_bar = $banner.find(".banner-progress-bar");
+    if ($progress_bar.length === 0) {
+        $progress_bar = $("<span>").addClass("banner-progress-bar");
+        $banner.prepend($progress_bar);
```
Prepending inside the banner root is a little awkward. Is there a reason the original insertion of the progress bar element shouldn't instead be an optional template feature?

sayamsamal replied:
Yeah we could do the progress bar as an optional template feature, I was trying to make the progress feature dynamic this way but I think nothing great is achieved doing that since it is just an empty span.

timabbott commented on web/src/navbar_alerts.ts:
```diff
@@ -283,12 +320,23 @@ export function initialize(): void {
+        unread_ops.mark_all_as_read();
+        setTimeout(() => {
+            banners.close($banner);
+        }, 2000);
```
So this looks like a basically fraudulent progress bar -- it just advances 25% every 300ms, while doing the actual operation in parallel.

I don't think that's a good idea. I'll see if you fix it later in the PR, but in general this sort of thing must be called out in the commit message.  I don't really see why this wasn't its own commit, actually.

timabbott replied:
Looks like this remains. I think probably we should just remove the progress bar from this PR, and plan to think about that as a follow-up? It seems not essential to the goals of this PR, and it might not be trivial to hook it up to the right data, so it should get its own development and review space aka PR.

sayamsamal replied:
Yep, the progress bar wasn't really an accurate depiction of the actual operation, but instead something to show for the 2000ms timeout before the operation is actually performed.

sayamsamal replied:
I have removed the progress bar and its application from the PR. Makes sense to implement it as a follow-up PR.

timabbott commented on web/src/navbar_alerts.ts:
```diff
@@ -190,6 +203,39 @@ const DESKTOP_NOTIFICATIONS_BANNER: AlertBanner = {
+        label: $t(
+            {
+                defaultMessage:
+                    "Welcome back! You have {old_unreads_missing, select, true {at least} other {}} {unread_msgs_count} unread messages. Do you want to mark them all as read?",
```
I'm not sure Transifex fully supports this i18n syntax. Generally I'd prefer to have the conditional logic visible in the code as an if/else statement before this `return` statement anyway.

sayamsamal replied:
I have removed this i18n syntax from the code, and replaced it with a conditional logic instead just before the `return` statement.

timabbott commented on web/src/navbar_alerts.ts:
```diff
@@ -250,18 +249,37 @@ const bankruptcy_banner = (): AlertBanner => {
+            $t_html(
+                {
+                    defaultMessage:
+                        "This <demo_link>demo organization</demo_link> will be automatically deleted in {days_remaining} days, unless it's <convert_link>converted into a permanent organization</convert_link>.",
```
These are better names than the old `z-general-link`, but do note that sort of change to strings require translators to re-translate the strings, which can be a lot of human work.

We also have a convention of using `<z-*link>` for these names. So `z-demo-link` and `z-convert-link` would be OK.

sayamsamal replied:
I'll keep that in mind, Changed the link translation tags to `z-demo-link` and `z-convert-link`.

timabbott commented on web/src/navbar_alerts.ts:
```diff
@@ -375,6 +385,10 @@ export function initialize(): void {
         );
     });
 
+    $("#navbar_alerts_wrapper").on("click", ".download_latest_zulip_version", () => {
```
You should use dashes for all the new class names like `download_latest_zulip_version` that you're introducing.

sayamsamal replied:
Oh, I was under the impression that we should be using snake_case for classes that are only being used for JS logic and don't have any CSS styling attached to them.

sayamsamal replied:
Updated the class names to use dashes instead of the snake case.

timabbott replied:
Nope, it's just a general naming convention preference, but we've not made any attempt to bulk-migrate them all, because that's likely to cause a lot of breakage.

timabbott commented on web/src/navbar_alerts.ts:
```diff
@@ -112,11 +111,23 @@ export function dismiss_upgrade_nag(ls: LocalStorage): void {
+export function is_organization_profile_incomplete(): boolean {
+    if (!current_user.is_admin) {
         return false;
     }
```
I think the `is_admin` check belongs in `should_show_organization_profile_incomplete_banner`, just to make the behavior match the function names.

sayamsamal replied:
Makes sense. Moved the `is_admin` check to `should_show_organization_profile_incomplete_banner` and updated the test cases.

timabbott commented on web/src/navbar_alerts.ts:
```diff
@@ -328,9 +328,9 @@ const demo_organization_deadline_banner = (): AlertBanner => {
+                        `<a class="banner-link" href="https://zulip.com/help/demo-organizations" target="_blank" rel="noopener noreferrer">${content_html.join("")}</a>`,
                     convert_link: (content_html) =>
-                        `<a class="banner__link" href="https://zulip.com/help/demo-organizations#convert-a-demo-organization-to-a-permanent-organization" target="_blank" rel="noopener noreferrer">${content_html.join("")}</a>`,
+                        `<a class="banner-link" href="https://zulip.com/help/demo-organizations#convert-a-demo-organization-to-a-permanent-organization" target="_blank" rel="noopener noreferrer">${content_html.join("")}</a>`,
```
I see `banner__link` still used in the CSS and showroom; is there a bug here?

timabbott commented on web/templates/left_sidebar.hbs:
```diff
@@ -143,6 +143,15 @@
+                    {{~!-- squash whitespace --~}}
+                    <span class="left-sidebar-navigation-label">{{t 'Change navbar banner' }}</span>
+                </a>
+            </li>
```
So this is cool and a thing I'd like to have available long term in some sort of "Zulip development tools" menu. We probably need to design such a menu in #design. But I think conceptually we want a menu that only appears if you've done something to trigger "developer mode" (say, a keyboard shortcut) and only works in the dev environment.

I assume we'll just not merge this commit when we merge the PR, so we're not blocked on figuring that out.

sayamsamal replied:
The idea of having **developer mode** sounds exciting to me. I'd be up for handling this proposal and setting one up.

timabbott commented:
OK, I did a full read of this and noted the above issues. I think it should be pretty easy to address them all in a fairly quick pass, since it's mostly missing translation tags. A testing technique that I recommend is using `manage.py makemessages` before/after the PR and reviewing a diff of the changes (might as well post it in a comment thread here). It should have some, since some things were replaced with buttons, but that'll be a pretty easy way to see that we didn't lose translations on strings unexpectedly.

sayamsamal commented:
Here is the output of the diff for the `translations.json`:
```diff
diff --git a/before_translations.json b/after_translations.json
index 6adef49..1c1d628 100644
--- a/before_translations.json
+++ b/after_translations.json
@@ -203 +202,0 @@
-  "Ask me later": "Ask me later",
@@ -328 +327 @@
-  "Complete your <z-link>organization profile</z-link>, which is displayed on your organization's registration and login pages.": "Complete your <z-link>organization profile</z-link>, which is displayed on your organization's registration and login pages.",
+  "Complete your organization profile, which is displayed on your organization's registration and login pages.": "Complete your organization profile, which is displayed on your organization's registration and login pages.",
@@ -332,0 +332 @@
+  "Configuration instructions": "Configuration instructions",
@@ -402,0 +403 @@
+  "Customize notifications": "Customize notifications",
@@ -503 +503,0 @@
-  "Do you want to mark them all as read?": "Do you want to mark them all as read?",
@@ -510 +510 @@
-  "Download the latest version.": "Download the latest version.",
+  "Download the latest version": "Download the latest version",
@@ -853 +852,0 @@
-  "Marking all messages as read…": "Marking all messages as read…",
@@ -1010 +1009 @@
-  "No, don't ask again": "No, don't ask again",
+  "No, don't ask again.": "No, don't ask again.",
@@ -1229 +1227,0 @@
-  "See how to configure email.": "See how to configure email.",
@@ -1370,0 +1369 @@
+  "This <z-demo-link>demo organization</z-demo-link> will be automatically deleted in {days_remaining} days, unless it's <z-convert-link>converted into a permanent organization</z-convert-link>.": "This <z-demo-link>demo organization</z-demo-link> will be automatically deleted in {days_remaining} days, unless it's <z-convert-link>converted into a permanent organization</z-convert-link>.",
@@ -1393 +1391,0 @@
-  "This is a <z-link-general>demo organization</z-link-general> and will be automatically deleted in {days_remaining} days, unless it's <z-link-convert>converted into a permanent organization</z-link-convert>.": "This is a <z-link-general>demo organization</z-link-general> and will be automatically deleted in {days_remaining} days, unless it's <z-link-convert>converted into a permanent organization</z-link-convert>.",
@@ -1574 +1571,0 @@
-  "We strongly recommend enabling desktop notifications. They help Zulip keep your team connected.": "We strongly recommend enabling desktop notifications. They help Zulip keep your team connected.",
@@ -1579 +1576,2 @@
-  "Welcome back!": "Welcome back!",
+  "Welcome back! You have at least {unread_msgs_count} unread messages. Do you want to mark them all as read?": "Welcome back! You have at least {unread_msgs_count} unread messages. Do you want to mark them all as read?",
+  "Welcome back! You have {unread_msgs_count} unread messages. Do you want to mark them all as read?": "Welcome back! You have {unread_msgs_count} unread messages. Do you want to mark them all as read?",
@@ -1623 +1620,0 @@
-  "Yes": "Yes",
@@ -1693 +1689,0 @@
-  "You have <z-link>at least {unread_msgs_count}</z-link> unread messages.": "You have <z-link>at least {unread_msgs_count}</z-link> unread messages.",
@@ -1695 +1690,0 @@
-  "You have <z-link>{unread_msgs_count}</z-link> unread messages.": "You have <z-link>{unread_msgs_count}</z-link> unread messages.",
@@ -1769 +1764 @@
-  "Zulip needs your permission to <z-link>enable desktop notifications.</z-link>": "Zulip needs your permission to <z-link>enable desktop notifications.</z-link>",
+  "Zulip needs your permission to enable desktop notifications for important messages.": "Zulip needs your permission to enable desktop notifications for important messages.",
```

Some of them have arose from the change in punctuations, is that okay? Or should I revert then back to simplify the translations?

sayamsamal commented:
@timabbott I have addressed the feedback and also commented the output of diff between the old and new `translations.json` files. Could you give this another look?

timabbott commented:
Nice, merged, thanks for all the work on this!

sayamsamal commented:
@timabbott the last commit of this PR was for testing purposes. Was merging that an intentional choice?

timabbott commented:
No, addressed via 55aa63425bba215e1baaa041e9b81457e95842eb (Seems better to avoid removing and re-adding the other code).
